// angular-i18next/esm2015/I18NextErrorHandlingStrategies.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function I18NextErrorHandlingStrategy() { }
if (false) {
    /**
     * @param {?} resolve
     * @param {?} reject
     * @return {?}
     */
    I18NextErrorHandlingStrategy.prototype.handle = function (resolve, reject) { };
}
export class NativeErrorHandlingStrategy {
    /**
     * @param {?} resolve
     * @param {?} reject
     * @return {?}
     */
    handle(resolve, reject) {
        return (/**
         * @param {?} err
         * @param {?=} t
         * @return {?}
         */
        (err, t) => {
            /** @type {?} */
            let result = {
                err: err,
                t: t
            };
            resolve(result);
        });
    }
}
export class StrictErrorHandlingStrategy {
    /**
     * @param {?} resolve
     * @param {?} reject
     * @return {?}
     */
    handle(resolve, reject) {
        return (/**
         * @param {?} err
         * @param {?=} t
         * @return {?}
         */
        (err, t) => {
            /** @type {?} */
            let result = {
                err: err,
                t: t
            };
            if (!err) {
                resolve(result);
                return;
            }
            reject(err);
        });
    }
}
