// angular-i18next/esm2015/ITranslationService.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function ITranslationService() { }
if (false) {
    /** @type {?} */
    ITranslationService.prototype.language;
    /** @type {?} */
    ITranslationService.prototype.languages;
    /** @type {?} */
    ITranslationService.prototype.events;
    /** @type {?} */
    ITranslationService.prototype.options;
    /**
     * @param {?} plugin
     * @return {?}
     */
    ITranslationService.prototype.use = function (plugin) { };
    /**
     * @param {?=} options
     * @return {?}
     */
    ITranslationService.prototype.init = function (options) { };
    /**
     * @param {?} key
     * @param {?=} options
     * @return {?}
     */
    ITranslationService.prototype.t = function (key, options) { };
    /**
     * @param {?} value
     * @param {?} format
     * @param {?} lng
     * @return {?}
     */
    ITranslationService.prototype.format = function (value, format, lng) { };
    /**
     * @param {?} key
     * @param {?} options
     * @return {?}
     */
    ITranslationService.prototype.exists = function (key, options) { };
    /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    ITranslationService.prototype.getFixedT = function (lng, ns) { };
    /**
     * @param {?} ns
     * @return {?}
     */
    ITranslationService.prototype.setDefaultNamespace = function (ns) { };
    /**
     * @param {?} lng
     * @return {?}
     */
    ITranslationService.prototype.dir = function (lng) { };
    /**
     * @param {?} lng
     * @return {?}
     */
    ITranslationService.prototype.changeLanguage = function (lng) { };
    /**
     * @param {?} namespaces
     * @return {?}
     */
    ITranslationService.prototype.loadNamespaces = function (namespaces) { };
    /**
     * @param {...?} params
     * @return {?}
     */
    ITranslationService.prototype.reloadResources = function (params) { };
    /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} key
     * @param {?} options
     * @return {?}
     */
    ITranslationService.prototype.getResource = function (lng, ns, key, options) { };
    /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} key
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    ITranslationService.prototype.addResource = function (lng, ns, key, value, options) { };
    /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} resources
     * @return {?}
     */
    ITranslationService.prototype.addResources = function (lng, ns, resources) { };
    /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} resources
     * @param {?} deep
     * @param {?} overwrite
     * @return {?}
     */
    ITranslationService.prototype.addResourceBundle = function (lng, ns, resources, deep, overwrite) { };
    /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    ITranslationService.prototype.hasResourceBundle = function (lng, ns) { };
    /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    ITranslationService.prototype.getResourceBundle = function (lng, ns) { };
    /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    ITranslationService.prototype.removeResourceBundle = function (lng, ns) { };
}
