// angular-i18next/esm5/I18NextEvents.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BehaviorSubject, Subject } from 'rxjs';
var I18NextEvents = /** @class */ (function () {
    function I18NextEvents() {
        this.initialized = new BehaviorSubject(false);
        this.loaded = new BehaviorSubject(false);
        this.failedLoading = new Subject();
        this.missingKey = new Subject();
        this.added = new Subject();
        this.removed = new Subject();
        this.languageChanged = new BehaviorSubject(null);
    }
    return I18NextEvents;
}());
export { I18NextEvents };
if (false) {
    /** @type {?} */
    I18NextEvents.prototype.initialized;
    /** @type {?} */
    I18NextEvents.prototype.loaded;
    /** @type {?} */
    I18NextEvents.prototype.failedLoading;
    /** @type {?} */
    I18NextEvents.prototype.missingKey;
    /** @type {?} */
    I18NextEvents.prototype.added;
    /** @type {?} */
    I18NextEvents.prototype.removed;
    /** @type {?} */
    I18NextEvents.prototype.languageChanged;
}
