// angular-i18next/esm5/I18NextService.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import * as i18n from 'i18next';
import { I18NEXT_ERROR_HANDLING_STRATEGY } from './I18NEXT_TOKENS';
import { I18NextEvents } from './I18NextEvents';
/** @type {?} */
var i18next = i18n.default;
var I18NextService = /** @class */ (function () {
    function I18NextService(errorHandlingStrategy) {
        this.errorHandlingStrategy = errorHandlingStrategy;
        this.events = new I18NextEvents();
        this.language = '';
        this.languages = [];
    }
    Object.defineProperty(I18NextService.prototype, "options", {
        get: /**
         * @return {?}
         */
        function () {
            return i18next.options;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} plugin
     * @return {THIS}
     */
    I18NextService.prototype.use = /**
     * @template THIS
     * @this {THIS}
     * @param {?} plugin
     * @return {THIS}
     */
    function (plugin) {
        i18next.use.call(i18next, plugin);
        return (/** @type {?} */ (this));
    };
    /**
     * @param {?=} options
     * @return {?}
     */
    I18NextService.prototype.init = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        var _this = this;
        options = options || {};
        this.subscribeEvents();
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            i18next.init.call(i18next, Object.assign({}, options), _this.errorHandlingStrategy.handle(resolve, reject));
        }));
    };
    /**
     * @param {?} key
     * @param {?=} options
     * @return {?}
     */
    I18NextService.prototype.t = /**
     * @param {?} key
     * @param {?=} options
     * @return {?}
     */
    function (key, options) {
        options = options || {};
        return i18next.t.call(i18next, (/** @type {?} */ (key)), options);
    };
    /**
     * @param {?} value
     * @param {?} format
     * @param {?} lng
     * @return {?}
     */
    I18NextService.prototype.format = /**
     * @param {?} value
     * @param {?} format
     * @param {?} lng
     * @return {?}
     */
    function (value, format, lng) {
        return i18next.format.call(i18next, value, format, lng);
    };
    /**
     * @param {?} key
     * @param {?} options
     * @return {?}
     */
    I18NextService.prototype.exists = /**
     * @param {?} key
     * @param {?} options
     * @return {?}
     */
    function (key, options) {
        return i18next.exists.call(i18next, key, options);
    };
    /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    I18NextService.prototype.getFixedT = /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    function (lng, ns) {
        return i18next.getFixedT.call(i18next, lng, ns);
    };
    /**
     * @param {?} ns
     * @return {?}
     */
    I18NextService.prototype.setDefaultNamespace = /**
     * @param {?} ns
     * @return {?}
     */
    function (ns) {
        i18next.setDefaultNamespace.call(i18next, ns);
    };
    /**
     * @param {?=} lng
     * @return {?}
     */
    I18NextService.prototype.dir = /**
     * @param {?=} lng
     * @return {?}
     */
    function (lng) {
        if (lng === void 0) { lng = undefined; }
        return i18next.dir.call(i18next, lng);
    };
    /**
     * @param {?} lng
     * @return {?}
     */
    I18NextService.prototype.changeLanguage = /**
     * @param {?} lng
     * @return {?}
     */
    function (lng) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            i18next.changeLanguage.call(i18next, lng, _this.errorHandlingStrategy.handle(resolve, reject));
        }));
    };
    /**
     * @param {?} namespaces
     * @return {?}
     */
    I18NextService.prototype.loadNamespaces = /**
     * @param {?} namespaces
     * @return {?}
     */
    function (namespaces) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            i18next.loadNamespaces.call(i18next, namespaces, _this.errorHandlingStrategy.handle(resolve, reject));
        }));
    };
    /**
     * @param {?} lngs
     * @param {?} callback
     * @return {?}
     */
    I18NextService.prototype.loadLanguages = /**
     * @param {?} lngs
     * @param {?} callback
     * @return {?}
     */
    function (lngs, callback) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            i18next.loadNamespaces.call(i18next, lngs, _this.errorHandlingStrategy.handle(resolve, reject));
        }));
    };
    //#region resource handling
    //#region resource handling
    /**
     * @param {...?} params
     * @return {?}
     */
    I18NextService.prototype.reloadResources = 
    //#region resource handling
    /**
     * @param {...?} params
     * @return {?}
     */
    function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
        i18next.reloadResources.apply(i18next, params);
    };
    /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} key
     * @param {?} options
     * @return {?}
     */
    I18NextService.prototype.getResource = /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} key
     * @param {?} options
     * @return {?}
     */
    function (lng, ns, key, options) {
        return i18next.getResource.call(i18next, lng, ns, key, options);
    };
    /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} key
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    I18NextService.prototype.addResource = /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} key
     * @param {?} value
     * @param {?} options
     * @return {?}
     */
    function (lng, ns, key, value, options) {
        i18next.addResource.call(i18next, lng, ns, key, value, options);
    };
    /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} resources
     * @return {?}
     */
    I18NextService.prototype.addResources = /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} resources
     * @return {?}
     */
    function (lng, ns, resources) {
        i18next.addResources.call(i18next, lng, ns, resources);
    };
    /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} resources
     * @param {?} deep
     * @param {?} overwrite
     * @return {?}
     */
    I18NextService.prototype.addResourceBundle = /**
     * @param {?} lng
     * @param {?} ns
     * @param {?} resources
     * @param {?} deep
     * @param {?} overwrite
     * @return {?}
     */
    function (lng, ns, resources, deep, overwrite) {
        i18next.addResourceBundle.call(i18next, lng, ns, resources, deep, overwrite);
    };
    /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    I18NextService.prototype.hasResourceBundle = /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    function (lng, ns) {
        return i18next.hasResourceBundle.call(i18next, lng, ns);
    };
    /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    I18NextService.prototype.getResourceBundle = /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    function (lng, ns) {
        return i18next.getResourceBundle.call(i18next, lng, ns);
    };
    /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    I18NextService.prototype.removeResourceBundle = /**
     * @param {?} lng
     * @param {?} ns
     * @return {?}
     */
    function (lng, ns) {
        i18next.removeResourceBundle.call(i18next, lng, ns);
    };
    //#endregion
    //#endregion
    /**
     * @private
     * @return {?}
     */
    I18NextService.prototype.subscribeEvents = 
    //#endregion
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        i18next.on.call(i18next, 'initialized', (/**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            _this.language = i18next.language;
            _this.languages = i18next.languages;
            _this.events.initialized.next(options);
        }));
        i18next.on.call(i18next, 'loaded', (/**
         * @param {?} loaded
         * @return {?}
         */
        function (loaded) { return _this.events.loaded.next(loaded); }));
        i18next.on.call(i18next, 'failedLoading', (/**
         * @param {?} lng
         * @param {?} ns
         * @param {?} msg
         * @return {?}
         */
        function (lng, ns, msg) { return _this.events.failedLoading.next({ lng: lng, ns: ns, msg: msg }); }));
        i18next.on.call(i18next, 'languageChanged', (/**
         * @param {?} lng
         * @return {?}
         */
        function (lng) {
            _this.language = i18next.language;
            _this.languages = i18next.languages;
            _this.events.languageChanged.next(lng);
        }));
        i18next.on.call(i18next, 'missingKey', (/**
         * @param {?} lngs
         * @param {?} namespace
         * @param {?} key
         * @param {?} res
         * @return {?}
         */
        function (lngs, namespace, key, res) { return _this.events.missingKey.next({ lngs: lngs, namespace: namespace, key: key, res: res }); }));
        i18next.on.call(i18next, 'added', (/**
         * @param {?} lng
         * @param {?} ns
         * @return {?}
         */
        function (lng, ns) { return _this.events.added.next({ lng: lng, ns: ns }); }));
        i18next.on.call(i18next, 'removed', (/**
         * @param {?} lng
         * @param {?} ns
         * @return {?}
         */
        function (lng, ns) { return _this.events.removed.next({ lng: lng, ns: ns }); }));
    };
    I18NextService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    I18NextService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [I18NEXT_ERROR_HANDLING_STRATEGY,] }] }
    ]; };
    return I18NextService;
}());
export { I18NextService };
if (false) {
    /** @type {?} */
    I18NextService.prototype.events;
    /** @type {?} */
    I18NextService.prototype.language;
    /** @type {?} */
    I18NextService.prototype.languages;
    /**
     * @type {?}
     * @private
     */
    I18NextService.prototype.errorHandlingStrategy;
}
